File: cpp_src/pi_control.cpp
// PIController.cpp

#include "pi_control.hpp"

PIController::PIController(double kp, double ki) : kp(kp), ki(ki), integral(0.0) {}

double PIController::compute_control(double reference, double measurement) {
    double error = reference - measurement;

    double control_out = kp * error + integral;
    integral           = ki * error + integral;

    return control_out;
}

File: rust_src/pi_control.rs
// rust_src/pi_controller.rs

pub struct PIController {
    pub kp: f64,        // Proportional gain
    pub ki: f64,        // Integral gain
    integral: f64,      // Integral term
}

impl PIController {
    pub fn new(kp: f64, ki: f64) -> Self {
        PIController {
            kp,
            ki,
            integral: 0.0,
        }
    }

    pub fn compute_control(&mut self, reference: f64, measurement: f64) -> f64 {
        let error = reference - measurement;
        let control_out = self.kp * error + self.integral;
        self.integral += self.ki * error;
        control_out
    }
}
